OpenTelemetry (S3M) Web Servers apps (#1795)
SumoLogic/sumologic-documentation: docs/integrations/web-servers/opentelemetry/apache-opentelemetry.md

### Sample Log Messages

Access Logs
```sh
192.168.29.177 - - [26/Apr/2021:12:18:32 +0530]  "GET /server-status HTTP/1.1"  404  196
```

Error Logs
```sh
[Mon Apr 26 09:52:58.188858 2021]  [core:notice]  [pid 530] AH00094: Command line: '/usr/sbin/httpd -D FOREGROUND'
```

### Sample Log Query

This sample Query is from the Top 5 Clients Causing 4xx Errors panel of the Apache - Web server Operations dashboard.

Query String

```sql
webengine.system=apache webengine.cluster.name=* HTTP (40* OR 41* OR 42* OR 43* OR 44* or 45* or 49*)
| json "log" nodrop | if  (_raw matches "{*", log, _raw)  as mesg
| parse regex field=mesg "^(?<src_ip>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})" nodrop
| parse regex field=mesg "(?<method>[A-Z]+)\s(?<url>\S+)\sHTTP\/[\d\.]+[\\n]*\"\s(?<status_code>\d+)\s(?<size>[\d-]+)" nodrop
| parse regex field=mesg "(?<method>[A-Z]+)\s(?<url>\S+)\sHTTP\/[\d\.]+[\\n]*\"\s(?<status_code>\d+)\s(?<size>[\d-]+)\s\"(?<referrer>.*?)\"\s\"(?<user_agent>.+?)\".*" nodrop
| where status_code matches "4*"
| count as count by src_ip
| sort count, src_ip asc
| limit  5
```

### Sample Metric 

```

> {"queryId":"A","server_name":"localhost","_source":"apache-otel-metric","state":"dnslookup","webengine.cluster":"apache_otel_cluster","sumo.datasource":"apache","deployment.environment":"user_apache_env","_contentType":"Carbon2","metric":"apache.scoreboard","_collectorId":"000000000C9802CD","_sourceId":"0000000042E3AAB3","webengine.system":"apache","unit":"{workers}","_collector":"Labs - apache-otel","max":100,"min":0,"avg":18.64,"sum":1100,"latest":1,"count":59}

Author: ```suggestion
{
	"queryId":"A",
	"server_name":"localhost",
	"_source":"apache-otel-metric",
	"state":"dnslookup",
	"webengine.cluster":"apache_otel_cluster",
	"sumo.datasource":"apache",
	"deployment.environment":"user_apache_env",
	"_contentType":"Carbon2",
	"metric":"apache.scoreboard",
	"_collectorId":"000000000C9802CD",
	"_sourceId":"0000000042E3AAB3",
	"webengine.system":"apache",
	"unit":"{workers}",
	"_collector":"Labs - apache-otel",
	"max":100,
	"min":0,
	"avg":18.64,
	"sum":1100,
	"latest":1,
	"count":59
}
```

The [Apache](https://httpd.apache.org/ABOUT_APACHE.html) app is a unified logs and metrics app that helps you monitor the availability, performance, health and resource utilization of Apache web server farms. Preconfigured dashboards and searches provide visibility into your environment for real-time or historical analysis: visitor locations, visitor access types, traffic patterns, errors, web server operations, resource utilization and access from known malicious sources.

The diagram below illustrates the components of the Apache collection for each web server. OpenTelemetry collector runs on the same host as Apache, and uses the [Apache Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/apachereceiver) to obtain Apache metrics, and the [Sumo Logic OpenTelemetry Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/sumologicexporter) to send the metrics to Sumo Logic. Apache logs are sent to Sumo Logic through a [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver).

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Apache-OpenTelemetry/Apache-Schematics.png')} alt="Schematics" />

Reviewer: This image link is broken when I view the page locally using Yarn.

The Sumo Logic App for Apache assumes:

-   The [NCSA extended/combined log file format ](http://httpd.apache.org/docs/current/mod/mod_log_config.html)has been configured for Apache access logs and the default error log format for Apache Access logs and Apache Error logs. For a list of metrics that are collected and used by the app, see [Apache Metrics](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/apachereceiver/documentation.md).

## Fields Creation in Sumo Logic for Apache

Following are the [Fields](https://help.sumologic.com/docs/manage/fields/) which will be created as part of Apache App install if not already present.

- **`webengine.cluster.name`** - User configured.Enter a name to uniquely identify your Apache web server cluster. This web server cluster name will be shown in the Sumo Logic dashboards.

> - **`webengine.system`** - Has fixed value of apache

Reviewer 2: ```suggestion
- **`webengine.system`** - Has fixed value of **apache**
```

> - **`sumo.datasource`** - Has fixed value of apache

Reviewer 2: ```suggestion
- **`sumo.datasource`** - Has fixed value of **apache**
```

> - **`webengine..node.name`** - Has the value of host name of the machine which is being monitored

Reviewer 2: ```suggestion
- **`webengine.node.name`** - Has the value of host name of the machine which is being monitored
```

## Prerequisites

> The receiver used, gets stats from an Apache Web Server instance using the server-status?auto endpoint.

Author: Put 'server-status?auto' in code font

```suggestion
The receiver used, gets stats from an Apache Web Server instance using the server-status?auto endpoint.
```

This receiver supports Apache Web Server version 2.4+

> In order to receive server statistics, you must configure the server's httpd.conf file to [enable status support.](https://httpd.apache.org/docs/2.4/mod/mod_status.html)

Reviewer: ```suggestion
In order to receive server statistics, you must configure the server's httpd.conf file to [enable status support.](https://httpd.apache.org/docs/2.4/mod/mod_status.html)
```
```suggestion
In order to receive server statistics, you must configure the server's `httpd.conf` file to [enable status support.](https://httpd.apache.org/docs/2.4/mod/mod_status.html)
```

> In order to receive server statistics, you must configure the server's httpd.conf file to [enable status support.](https://httpd.apache.org/docs/2.4/mod/mod_status.html)

Reviewer: ```suggestion
In order to receive server statistics, you must configure the server's `httpd.conf` file to [enable status support.](https://httpd.apache.org/docs/2.4/mod/mod_status.html)
```

Configure Apache to log to a local files. Before you can configure Sumo Logic to ingest logs, you must configure the logging of access logs and error logs via the instructions described in their [documentation](https://httpd.apache.org/docs/2.4/logs.html).

> To configure the Apache log files, locate your local `httpd.conf` configuration file in the Apache directory. After determining the location of the conf file, modify the httpd.conf configuration file logging parameters if required.

Reviewer: ```suggestion
To configure the Apache log files, locate your local `httpd.conf` configuration file in the Apache directory. After determining the location of the conf file, modify the `httpd.conf` configuration file logging parameters if required.
```

For access logs, the following directive is to be noted:
-   CustomLog: access log file path and format (standard common and combined)

For error logs, following directives are to be noted:
-   ErrorLog: error log file path
-   LogLevel: to control the number of messages logged to the `error_log`

## Collecting Logs, Metrics & Installing App for Apache

Here are the steps for Collecting Logs, metric and installing the app:

### Step 1: Set up Collector

:::note

> If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the "Use an existing Collector" option.

Author: Do not use quotes for interface elements - please change to bold

```suggestion
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the "Use an existing Collector" option.
```

:::

> If you want to use an existing OpenTelemetry Collector then this step can be skipped by selecting the option of using an existing Collector.

Reviewer 2: This line is repeated. The one with the note section should exist and this line can be deleted.

> If you want to use an existing OpenTelemetry Collector then this step can be skipped by selecting the option of using an existing Collector.

Reviewer: ```suggestion

```

> If you want to create a new Collector please select **Add a new Collector** option.

Reviewer 2: To create a new Collector:

1. Select the Add a new Collector option.

2. Select the platform for which you want to install the Sumo OpenTelemetry Collector.

Select the platform for which you want to install the Sumo OpenTelemetry Collector.

This will generate a command that can be executed in the machine that needs to get monitored. Once executed, it will install the Sumo Logic OpenTelemetry Collector agent.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Apache-OpenTelemetry/Apache-Collector.png')} alt="Collector" />

Reviewer: This image link is broken.

### Step 2: Configure integration

> In this step we will be configuring the yaml required for Apache Collection.

Author: ```suggestion
In this step, we will be configuring the yaml required for Apache Collection.
```

> Below are the input required :

Author: ```suggestion
Below are the input required:
```

> Below are the input required:

Reviewer: ```suggestion
Below are the inputs required:
```

> -   **Endpoint** - (default: http://localhost:8080/server-status?auto): The URL of the httpd status endpoint

Reviewer: ```suggestion
-   **Endpoint** - (default: `http://localhost:8080/server-status?auto`): The URL of the httpd status endpoint
```

-   **Access File log Path** - Enter the path to the Access log file for your mysql instance.
-   **Error file log path** - Enter the path to the error log file for your mysql instance.
-   Fields - **webengine.cluster.name**

> 

Reviewer 2: ```suggestion
You can add any custom fields which you want to tag along with the data ingested in sumo.
```

Click on the **Download YAML File** button to get the yaml file.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Apache-OpenTelemetry/Apache-YAML.png')} alt="YAML" />

Reviewer: This image link is broken.

> ### Step 3: Sending logs and metric to Sumo

Reviewer: ```suggestion
### Step 3: Sending logs and metrics to Sumo
```

Once you download the yaml file, please follow the below steps based on your platform.

<Tabs
  className="unique-tabs"
  defaultValue="Linux"
  values={[
    {label: 'Linux', value: 'Linux'},
    {label: 'Windows', value: 'Windows'},

>     {label: 'macOS', value: 'macOS'},

Reviewer: ```suggestion
    {label: 'Mac', value: 'Mac'},
```

Author: I'm going to stet these to maintain consistency with the rest of our docs, which all use macOS. Definitely open to discussing at a team meeting!

  ]}>

<TabItem value="Linux">

1.  Copy the yaml to /etc/otelcol-sumo/conf.d/ folder for the Apache instance which needs to be monitored.

> 2.  restart the collector using

Reviewer: ```suggestion
2.  Restart the collector using:
```

```sh
 sudo systemctl restart otelcol-sumo
```

</TabItem>
<TabItem value="Windows">

1.  Copy the yaml to `C:\ProgramData\Sumo Logic\OpenTelemetry Collector\config\conf.d` folder in the machine which needs to be monitored.
2.  Restart the collector using 
```sh
Restart-Service -Name OtelcolSumo
```

</TabItem>

> <TabItem value="macOS">

Author: https://github.com/SumoLogic/sumologic-documentation/pull/1795/files#r1142949013

1.  Copy the yaml to /etc/otelcol-sumo/conf.d/ folder in the Apache instance which needs to be monitored.

> 2.  Restart the otelcol-sumo process using the below command 

Reviewer: ```suggestion
2.  Restart the otelcol-sumo process using the below command:
```

```sh
 otelcol-sumo --config /etc/otelcol-sumo/sumologic.yaml --config "glob:/etc/otelcol-sumo/conf.d/*.yaml"
```

</TabItem>
</Tabs>

After successful execution of the above command, Sumo will start receiving the data from your host machine.

> Press Next. This will install the app to your Sumo Logic Org. The app consists of Dashboard and monitors

Reviewer 2: ```suggestion
Press **Next**. This will install the app to your Sumo Logic Org. The app consists of Dashboard and monitors
```

> Press **Next**. This will install the app to your Sumo Logic Org. The app consists of Dashboard and monitors

Reviewer: ```suggestion
Press **Next**. This will install the app to your Sumo Logic Org. The app consists of a Dashboard and monitors.
```

Panels will start to fill automatically. It's important to note that each panel slowly fills with data matching the time range query and received since the panel was created. Results won't immediately be available, but within 20 minutes you'll see full graphs and maps.

### Sample Log Messages

Access Logs
```sh
192.168.29.177 - - [26/Apr/2021:12:18:32 +0530]  "GET /server-status HTTP/1.1"  404  196
```

Error Logs
```sh
[Mon Apr 26 09:52:58.188858 2021]  [core:notice]  [pid 530] AH00094: Command line: '/usr/sbin/httpd -D FOREGROUND'
```

### Sample Log Query

This sample Query is from the Top 5 Clients Causing 4xx Errors panel of the Apache - Web server Operations dashboard.

Query String

```sql

> webengine.system=apache webengine.cluster=* HTTP (40* OR 41* OR 42* OR 43* OR 44* or 45* or 49*)

Reviewer 2: ```suggestion
webengine.system=apache webengine.cluster.name=* HTTP (40* OR 41* OR 42* OR 43* OR 44* or 45* or 49*)
```

| json "log" nodrop | if  (_raw matches "{*", log, _raw)  as mesg
| parse regex field=mesg "^(?<src_ip>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})" nodrop
| parse regex field=mesg "(?<method>[A-Z]+)\s(?<url>\S+)\sHTTP\/[\d\.]+[\\n]*\"\s(?<status_code>\d+)\s(?<size>[\d-]+)" nodrop
| parse regex field=mesg "(?<method>[A-Z]+)\s(?<url>\S+)\sHTTP\/[\d\.]+[\\n]*\"\s(?<status_code>\d+)\s(?<size>[\d-]+)\s\"(?<referrer>.*?)\"\s\"(?<user_agent>.+?)\".*" nodrop
| where status_code matches "4*"
| count as count by src_ip
| sort count, src_ip asc
| limit  5
```

### Sample Metric 

```
{"queryId":"A","server_name":"localhost","_source":"apache-otel-metric","state":"dnslookup","webengine.cluster":"apache_otel_cluster","sumo.datasource":"apache","deployment.environment":"user_apache_env","_contentType":"Carbon2","metric":"apache.scoreboard","_collectorId":"000000000C9802CD","_sourceId":"0000000042E3AAB3","webengine.system":"apache","unit":"{workers}","_collector":"Labs - apache-otel","max":100,"min":0,"avg":18.64,"sum":1100,"latest":1,"count":59}
```

### Sample Metric query

Sample query from **Waiting** panel in Request State Analysis dashboard

```

> sumo.datasource=apache metric=apache.scoreboard state=waiting webengine.cluster=* server_name=* | sum by webengine.cluster

Reviewer 2: ```suggestion
sumo.datasource=apache metric=apache.scoreboard state=waiting webengine.cluster.name=* server_name=* | sum by webengine.cluster
```

Configure Apache to log to a local files. Before you can configure Sumo Logic to ingest logs, you must configure the logging of access logs and error logs via the instructions described in their [documentation](https://httpd.apache.org/docs/2.4/logs.html).

To configure the Apache log files, locate your local `httpd.conf` configuration file in the Apache directory. After determining the location of the conf file, modify the `httpd.conf` configuration file logging parameters if required.

For access logs, the following directive is to be noted:
-   CustomLog: access log file path and format (standard common and combined)

For error logs, following directives are to be noted:
-   ErrorLog: error log file path
-   LogLevel: to control the number of messages logged to the `error_log`

## Collecting Logs, Metrics & Installing App for Apache

Here are the steps for Collecting Logs, metric and installing the app:

### Step 1: Set up Collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the **Use an existing Collector** option.
:::

> If you want to create a new Collector please select **Add a new Collector** option.

Author: ```suggestion
To create a new Collector:

1. Select the Add a new Collector option.
1. Select the platform for which you want to install the Sumo Logic OpenTelemetry Collector.
```

| json "log" nodrop | if  (_raw matches "{*", log, _raw)  as mesg
| parse regex field=mesg "^(?<src_ip>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})" nodrop
| parse regex field=mesg "(?<method>[A-Z]+)\s(?<url>\S+)\sHTTP\/[\d\.]+[\\n]*\"\s(?<status_code>\d+)\s(?<size>[\d-]+)" nodrop
| parse regex field=mesg "(?<method>[A-Z]+)\s(?<url>\S+)\sHTTP\/[\d\.]+[\\n]*\"\s(?<status_code>\d+)\s(?<size>[\d-]+)\s\"(?<referrer>.*?)\"\s\"(?<user_agent>.+?)\".*" nodrop
| where status_code matches "4*"
| count as count by src_ip
| sort count, src_ip asc
| limit  5
```

### Sample Metric 

```
{
	"queryId":"A",
	"server_name":"localhost",
	"_source":"apache-otel-metric",
	"state":"dnslookup",
	"webengine.cluster":"apache_otel_cluster",
	"sumo.datasource":"apache",
	"deployment.environment":"user_apache_env",
	"_contentType":"Carbon2",
	"metric":"apache.scoreboard",
	"_collectorId":"000000000C9802CD",
	"_sourceId":"0000000042E3AAB3",
	"webengine.system":"apache",
	"unit":"{workers}",
	"_collector":"Labs - apache-otel",
	"max":100,
	"min":0,
	"avg":18.64,
	"sum":1100,
	"latest":1,
	"count":59
}
```

### Sample Metric query

> Sample query from **Waiting** panel in **Request State Analysis** dashboard

Author: ```suggestion
Here's a sample query from the **Request State Analysis** dashboard > **Waiting** panel:
```

### Visitor Locations

The **Apache - Visitor Locations** dashboard provides a high-level view of Apache visitor geographic locations both worldwide and in the United States.

Use this dashboard to:

-   Get insights into geographic locations of your user base

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Apache-OpenTelemetry/Apache-Visitor-Locations.png')} alt="Visitor Locations" />

### Visitor Access Types

The **Apache - Visitor Access Types** dashboard provides insights into visitor platform types, browsers, device types, and operating systems.

Use this dashboard to:

-   Understand which platform and browsers are being used to access your applications

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Apache-OpenTelemetry/Apache-Visitor-Access-Types.png')} alt="Visitor Access Types" />

### Visitor Traffic Insight

The **Apache - Visitor Traffic Insight** dashboard provides summarized information on the top URLs, referrers, search terms, and media types served.

Use this dashboard to:

> -   To understand content types of content that are frequently requested by users.

Author: ```suggestion
-   Understand content types of content that are frequently requested by users.
```

To configure the Apache log files, locate your local `httpd.conf` configuration file in the Apache directory. After determining the location of the conf file, modify the `httpd.conf` configuration file logging parameters if required.

For access logs, the following directive is to be noted:
-   CustomLog: access log file path and format (standard common and combined)

For error logs, following directives are to be noted:
-   ErrorLog: error log file path
-   LogLevel: to control the number of messages logged to the `error_log`

## Collecting Logs, Metrics & Installing App for Apache

Here are the steps for Collecting Logs, metric and installing the app:

### Step 1: Set up Collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the **Use an existing Collector** option.
:::

To create a new Collector:

1. Select the **Add a new Collector** option.
1. Select the platform for which you want to install the Sumo Logic OpenTelemetry Collector.

> Select the platform for which you want to install the Sumo OpenTelemetry Collector.

Author: ```suggestion
```
